### PR TITLE
Implement a server res directory for server required client files

### PR DESCRIPTION
**Pull request by EmosewaMC**
Tested that a sqlite file in the res path is correctly copied to `resServer`
Tested that if no sqlite file exists in the client res directory, a new one is created in `resServer`
Tested that is the sqlite file exists in resServer but not in the client res path, the file is not converted again

### CODE REVIEW

**Comment by HailStorm32** [comment]
The README will also need to be updated if you are renaming the res folder

**Comment by EmosewaMC** [comment]
> The README will also need to be updated if you are renaming the res folder

Waiting on the other readme update and if that is merged first ill update it there.

**Comment by Jettford** [comment]
This will make simply dragging and dropping binaries into a client in the future more difficult, I do not agree with this change however half of that is because there is no reason for these changes listed here. Please outline reasons for this change.

**Comment by luxaritas** [comment]
I pushed for this. Best practice for daemons is that they only write to directories which do not contain any other content aside from what it contains. The rationale is twofold:
1) Being intentional about not modifying input file locations. The assumption is that the client path is being used solely to read from, and having files written there. If the server needs to create additional files, they should be created somewhere it "owns", unless the explicit purpose is to generate those files in the manner the user asked.
2) Server hardening. As a manner of best practice, daemons should only have write access to what they strictly need to. Currently, you need to allow writes into build and res, which could allow for unintended writes of application files given another vulnerability. This provides defense-in-depth.

Regardless, I don't see why this has to make shipping binaries harder. The server can default create serverRes/CDServer.SQLite (including the parent dir) as well as it can res/CDServer.sqlite

**Comment by Jettford** [comment]
You are missing the point, if you were to say drop the binaries into a client the res is called res, it won't know where to source the cdclient.fdb to convert.

**Comment by Jettford** [comment]
Ok after re-going through the source I think I actually understand now but this just raises a different question of why does it matter if the server has R and W on client files?

**Comment by luxaritas** [comment]
It's not critical, but I still maintain that as a matter of practice we shouldn't be modifying someone's client in any way. The CDServer is a server file. Keep it in a server-controlled area.

**Review by aronwk-aaron — APPROVED** [review]

**Comment by EmosewaMC** [comment]
> The README will also need to be updated if you are renaming the res folder

This is not needed anymore since no user needs to interact with a server res directory with the new readme structure

File: dMasterServer/MasterServer.cpp
Comment (Xiphoseer) [inline dMasterServer/MasterServer.cpp:147]: might make sense to extract those exists calls into variables if you use them multiple times.

File: dMasterServer/MasterServer.cpp
Comment (Xiphoseer) [inline dMasterServer/MasterServer.cpp:162]: I think this condition is a bit weird to reason about. Yes, if both don't exist, then it would have gone to the first if case, so if the sqlite in resServer does not exist, and we get here there must be a sqlite in res to move over.

But considering, that this is essentially a migration, I think this should just be a positive check for existance, nested under a simple "resServer sqlite missing" before converting from fdb

File: dMasterServer/MasterServer.cpp
Comment (Xiphoseer) [inline dMasterServer/MasterServer.cpp:165]: Add a comment that this copies instead of move because the folder may be read only

**Review by Xiphoseer — DISMISSED** [review]

File: dMasterServer/MasterServer.cpp
Comment (EmosewaMC) [reply dMasterServer/MasterServer.cpp:165]: Thats fair - ill add those

**Comment by EmosewaMC** [comment]
Tested having fdb and no sqlite, having fdb and sqlite in the old directory, having sqlite in new directory with packed and unpacked clients.  All resulted in cdserver being in resServer

**Review by aronwk-aaron — APPROVED** [review]

File: dMasterServer/MasterServer.cpp
Comment (Xiphoseer) [inline dMasterServer/MasterServer.cpp:151]: Can you swap the branches and remove the `!`?

**Review by Xiphoseer — APPROVED** [review]

File: dMasterServer/MasterServer.cpp
Comment (EmosewaMC) [reply dMasterServer/MasterServer.cpp:151]: yes - will do so in the morning

**Review by Xiphoseer — DISMISSED** [review]
LGTM

File: dMasterServer/MasterServer.cpp
Comment (EmosewaMC) [reply dMasterServer/MasterServer.cpp:171]: Mistake on my end.  Its supposed to return FinalizeShutdown(); so destroy all objects that exist

**Review by aronwk-aaron — APPROVED** [review]

**Review by Xiphoseer — DISMISSED** [review]

**Review by aronwk-aaron — APPROVED** [review]

**Review by Xiphoseer — APPROVED** [review]

**Pull request merged**
